slotsmanager.c

```diff
@@ -0,0 +1,427 @@
+
+            memcpy(BASESLOT_PI(i, ptype),
+                BASESLOT_PI(used - 1, ptype), elsize);
+            used -- ;
+        }
+        i ++;
+    }
+
+    SlotsManager->info[ptype].size = used;
+}
+
+/* defrags ensures locality. */
+static void
+slots_gc_defrag(int ptype)
+{
+
+    if(!SlotsManager->info[ptype].enabled) return;
+    /* measure the fragmentation */
+    int i;
+    int frag = 0;
+    for(i = 1;
+        i < SlotsManager->info[ptype].size;
+        i ++) {
+
+        if( BASESLOT_PI(i, ptype)->gc.ReverseLink <
+            BASESLOT_PI(i - 1, ptype)->gc.ReverseLink
+            ) {
+            frag++;
+        }
+    }
+    int defrag = MPIU_Any(
+            frag > SlotsManager->info[ptype].size * 0.1,
+            MPI_COMM_WORLD);
+
+    /* if any rank is too fragmented, add a sorting to defrag. */
+    if(defrag)  {
+        /* sort the used ones
+         * by their location in the P array */
+        qsort_openmp(SlotsManager->info[ptype].ptr,
+                     SlotsManager->info[ptype].size,
+                     SlotsManager->info[ptype].elsize, 
+                     slot_cmp_reverse_link);
+    }
+}
+
+/* update new pointers. */
+static void
+slots_gc_collect(int ptype)
+{
+    int i;
+    if(!SlotsManager->info[ptype].enabled) return;
+
+    /* Now update the link in BhP */
+#pragma omp parallel for
+    for(i = 0;
+        i < SlotsManager->info[ptype].size;
+        i ++) {
+
+        P[BASESLOT_PI(i, ptype)->gc.ReverseLink].PI = i;
+    }
+}
+
+static int
+slots_gc_slots()
+{
+    int ptype;
+
+    int64_t total0[6];
+    int64_t total1[6];
+
+    int disabled = 1;
+    for(ptype = 0; ptype < 6; ptype ++) {
+        sumup_large_ints(1, &SlotsManager->info[ptype].size, &total0[ptype]);
+        if(total0[ptype] != 0) disabled = 0;
+    }
+    /* disabled this if no slots are used */
+    if(disabled) {
+        return 0;
+    }
+
+    slots_gc_mark();
+
+    for(ptype = 0; ptype < 6; ptype++) {
+        slots_gc_sweep(ptype);
+        slots_gc_defrag(ptype);
+        slots_gc_collect(ptype);
+    }
+
+#ifdef DEBUG
+    slots_check_id_consistency();
+#endif
+
+    for(ptype = 0; ptype < 6; ptype ++) {
+        sumup_large_ints(1, &SlotsManager->info[ptype].size, &total1[ptype]);
+
+        if(total1[ptype] != total0[ptype])
+            message(0, "GC: Reducing number of slots for %d from %ld to %ld\n", ptype, total0[ptype], total1[ptype]);
+    }
+
+    /* slot gc never invalidates the tree */
+    return 0;
+}
+
+
+void
+slots_reserve(int atleast[6])
+{
+    int newMaxSlots[6];
+    int ptype;
+    int good = 1;
+
+    if(SlotsManager->Base == NULL)
+        SlotsManager->Base = (char*) mymalloc("SlotsBase", 0);
+
+    for(ptype = 0; ptype < 6; ptype ++) {
+        newMaxSlots[ptype] = SlotsManager->info[ptype].maxsize;
+        while(newMaxSlots[ptype] < atleast[ptype]) {
+            int add = 0.2 * newMaxSlots[ptype];
+            if (add < 128) add = 128;
```

Comment:
You have changed the increase amount here (made it smaller). I am worried that the new condition does not ensure that there are enough slots available for the next step's star formation, if, e.g. newMaxSlots = 1.21 * atleast. (although I agree that the old condition was overly complex, given that the SPH slot needs increasing very very rarely).

I think a fix would be to scale this with total particle load: add = 0.005 * All.MaxPart. That way we don't risk running out of slots in highly threaded situations.

Reply:
Is 0.5% enough if we have very fast SFR? This is probably point to the need to grow in fork.

Reply:
Yeah, I don't know, but it's probably ok. The last version I had increased the star allocation every time N_star_slots > MaxPartStar - 0.005*MaxPart, so it had to exceed 0.5% of the particles in every timestep, which seemed pretty safe? I would guess that the timestep would shrink if star formation got faster than that.

Reply:
So We'll always set MaxSlots to 0.5 % + the requested reservation.

Reply:
That's not quite as aggressive as it was before: before it increased the allocation if N_star_slots was within 0.5% of the total. Now it is possible for N_star_slots to get arbitrarily close to the total before being increased. If a burst of star formation occurs at this point we are in trouble.